### PR TITLE
[JIT] ARM64 - Do not allow move and shifting with MSL on 16-bit vectors

### LINKED ISSUE #76880: Test failure: JIT\\Regression\\JitBlue\\GitHub_23159\\GitHub_23159\\GitHub_23159.cmd
win-arm64 random JitStress

https://dev.azure.com/dnceng-public/public/_build/results?buildId=47392&view=ms.vss-test-web.build-test-results-tab&runId=954268&resultId=102958&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

```
set COMPlus_TieredCompilation=0
set COMPlus_JitStress=e7

    JIT\Regression\JitBlue\GitHub_23159\GitHub_23159\GitHub_23159.cmd [FAIL]
      
      Return code:      1
      Raw output file:      D:\h\w\9EF6090D\w\AEE409BF\uploads\Reports\JIT.Regression\JitBlue\GitHub_23159\GitHub_23159\GitHub_23159.output.txt
      Raw output:
      BEGIN EXECUTION
       "D:\h\w\9EF6090D\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false"  GitHub_23159.dll 
      Expected: 100
      Actual: -1
      END EXECUTION - FAILED
```

fwiw, this test has a lot of `Vector<T>` and `Unsafe` code

@dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by TIHan**
**Decription**

Should resolve: https://github.com/dotnet/runtime/issues/76880

We were allowing the `movi` instruction in conjunction with `MSL` to be emitted on 16-bit vectors which is not supported; it is only supported for 32-bit vectors. See https://developer.arm.com/documentation/ddi0596/2021-12/SIMD-FP-Instructions/MOVI--Move-Immediate--vector--?lang=en

**Acceptance Criteria**
- [x] JIT Stress run passes

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
Cool, thanks! What code do we generate instead?

**Comment by TIHan** [comment]
We generate:
```
mvni    v16.8h, #0
```
Which basically does an inverse of the bits if I'm reading this correctly: https://developer.arm.com/documentation/ddi0596/2021-12/SIMD-FP-Instructions/MVNI--Move-inverted-Immediate--vector--?lang=en

This is determined in the emitter from this logic:
```cpp
                // Next try the ones-complement form of the 'immediate' imm(i8,bySh)
                if ((elemsize == EA_2BYTE) || (elemsize == EA_4BYTE)) // Only EA_2BYTE or EA_4BYTE forms
                {
                    notOfImm  = NOT_helper(imm, getBitWidth(elemsize));
                    canEncode = canEncodeByteShiftedImm(notOfImm, elemsize, true, &bsi);
                    if (canEncode)
                    {
                        imm = bsi.immBSVal;
                        ins = INS_mvni; // uses a mvni encoding
                        assert(isValidImmBSVal(imm, size));
                        fmt = IF_DV_1B;
                        break;
                    }
                }
```

**Comment by AndyAyersMS** [comment]
Wonder if something bad happened to our azure storage:
```
call powershell Invoke-WebRequest -Uri "[https://clrjit.blob.core.windows.net/clang-tools/windows/clang-format.exe"](https://clrjit.blob.core.windows.net/clang-tools/windows/clang-format.exe%22) -OutFile bin\clang-format.exe
Invoke-WebRequest : ResourceNotFoundThe specified resource does not exist.
```
Similar problems in the SPMI legs.

@bruceforstall any ideas?

**Comment by BruceForstall** [comment]
The Azure Storage issue should be fixed now. Please re-run the failing jobs.

**Comment by AndyAyersMS** [comment]
@TIHan can this happen in .NET 7 or is this something new? Wondering if we might need to backport.

**Comment by TIHan** [comment]
@AndyAyersMS It looks like it could happen in .NET 7. Some of the code surrounding this hasn't changed since 2015.

**Comment by AndyAyersMS** [comment]
The simple repro you created in #76880 works ok on .NET 7 RC1, so perhaps this is a recent change? Also, the test that fails here probably hasn't been failing for long. Might be good to pin down when our codegen changed.

```
; Assembly listing for method X:Main(<unnamed>):int
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; No PGO data

G_M000_IG01:                ;; offset=0000H
        A9BE7BFD          stp     fp, lr, [sp,#-0x20]!
        6D0127E8          stp     d8, d9, [sp,#0x10]
        910003FD          mov     fp, sp

G_M000_IG02:                ;; offset=000CH
        12800000          movn    w0, #0
        4E040C10          dup     v16.4s, w0
        4EB01E08          mov     v8.16b, v16.16b
        D29EC100          movz    x0, #0xF608
        F2B4D580          movk    x0, #0xA6AC LSL #16
        F2CFFFA0          movk    x0, #0x7FFD LSL #32
        6E084509          mov     v9.d[0], v8.d[1]
        97FFFCE4          bl      CORINFO_HELP_NEWSFAST
        6E180528          mov     v8.d[1], v9.d[0]
        3C808008          str     q8, [x0,#0x08]
        D28F5601          movz    x1, #0x7AB0
        F2B4D661          movk    x1, #0xA6B3 LSL #16
        F2CFFFA1          movk    x1, #0x7FFD LSL #32
        F9400021          ldr     x1, [x1]
        D63F0020          blr     x1
        2A1F03E0          mov     w0, wzr

G_M000_IG03:                ;; offset=004CH
        6D4127E8          ldp     d8, d9, [sp,#0x10]
        A8C27BFD          ldp     fp, lr, [sp],#0x20
        D65F03C0          ret     lr

; Total bytes of code 88

<65535, 65535, 65535, 65535, 65535, 65535, 65535, 65535>
```

**Comment by TIHan** [comment]
The disasm you posted is optimized. I think it only emits movi when it is not.

**Comment by AndyAyersMS** [comment]
> The disasm you posted is optimized. I think it only emits movi when it is not.

True -- fixed it above. Still doesn't have the bug.

**Comment by TIHan** [comment]
I'll try to investigate when this started occurring then.

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib any way to get past this without starting from scratch?

```
Artifact SuperPMI_Asmdiffs_x86_checked already exists for build 55025.
```

**Comment by BruceForstall** [comment]
> @dotnet/jit-contrib any way to get past this without starting from scratch?
> 
> ```
> Artifact SuperPMI_Asmdiffs_x86_checked already exists for build 55025.
> ```

You can ignore it. This happens we using "Rerun" to re-run a portion of a job. We use the AzDO `PublishPipelineArtifact@1` task to upload log files, and it doesn't allow overwriting previously uploaded log files. Most of the YAML files in dotnet/runtime use the `continueOnError: true` option to simply ignore any failures. The JIT ones (currently) don't do this, so we don't ignore the failures, but we also use `condition: always()` on all subsequent tasks so we don't stop when hitting this failure, either. Even if we did use `continueOnError: true`, a rerun still wouldn't overwrite the old log file, which seems like odd behavior.

There is some discussion on this here, for example: https://github.com/MicrosoftDocs/azure-devops-docs/issues/10543

**Comment by markples** [comment]
For logs, using $(System.JobAttempt) is probably a good fit.  However, I don't understand why the docs say "Available in templates? -- no", so it might take some variable shenanigans to get the value in.

**Comment by TIHan** [comment]
I'll run a local SuperPMI diff to see what we get.

**Comment by TIHan** [comment]
There are no SuperPMI diffs found:
```
[12:23:39] ================ Logging to c:\work\runtime\artifacts\spmi\superpmi.6.log
[12:23:39] Using JIT/EE Version from jiteeversionguid.h: 982ed1fd-7bf3-425b-8b8a-902873151e79
[12:23:39] Using coredistools found at c:\work\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\coredistools.dll
[12:23:39] Found download cache directory "c:\work\runtime\artifacts\spmi\mch\982ed1fd-7bf3-425b-8b8a-902873151e79.windows.arm64" and --force_download not set; skipping download
[12:23:39] SuperPMI ASM diffs
[12:23:39] Base JIT Path: C:\work\sandbox\runtime\artifacts\bin\coreclr\windows.x64.Checked\clrjit_universal_arm64_x64.dll
[12:23:39] Diff JIT Path: c:\work\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\clrjit_universal_arm64_x64.dll
[12:23:39] Using MCH files:
[12:23:39]   c:\work\runtime\artifacts\spmi\mch\982ed1fd-7bf3-425b-8b8a-902873151e79.windows.arm64\benchmarks.run.windows.arm64.checked.mch
[12:23:39]   c:\work\runtime\artifacts\spmi\mch\982ed1fd-7bf3-425b-8b8a-902873151e79.windows.arm64\libraries.crossgen2.windows.arm64.checked.mch
[12:23:39] Running asm diffs of c:\work\runtime\artifacts\spmi\mch\982ed1fd-7bf3-425b-8b8a-902873151e79.windows.arm64\benchmarks.run.windows.arm64.checked.mch
[12:23:43] SuperPMI encountered missing data. Missing with base JIT: 47. Missing with diff JIT: 47. Total contexts: 34133.
[12:23:43] Running asm diffs of c:\work\runtime\artifacts\spmi\mch\982ed1fd-7bf3-425b-8b8a-902873151e79.windows.arm64\libraries.crossgen2.windows.arm64.checked.mch
[12:23:53] SuperPMI encountered missing data. Missing with base JIT: 55. Missing with diff JIT: 55. Total contexts: 238787.
[12:23:53] Asm diffs summary:
[12:23:53]   Summary Markdown file: c:\work\runtime\artifacts\spmi\diff_summary.2.md
[12:23:53]   No asm diffs
```

File: src/coreclr/jit/emitarm64.cpp [hunk 1]
```diff
@@ -3041,7 +3041,9 @@ emitter::code_t emitter::emitInsCode(instruction ins, insFormat fmt)
             {
                 canEncode = true;
             }
-            if (allow_MSL)
+
+            // MSL is only supported for 32-bit.
+            if (allow_MSL && size == EA_4BYTE)
```
Comment (BruceForstall) [inline src/coreclr/jit/emitarm64.cpp:3046]: nit: coding conventions specify fully-parenthesized expressions
```suggestion
            if (allow_MSL && (size == EA_4BYTE))
```

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**